ADR/0035-artifact-streams.md

```diff
@@ -0,0 +1,150 @@
+
+This model presented multiple challenges including:
+
+- If a user wants to build a new branch of an already onboarded git repository, they would need to create
+  a new [Component] (with a workspace-unique name).
+- [Component]s can be members of only one [Application].
+- As the number of related [Component]s increases, it becomes much harder to visualize all of the currently
+  or previously running PipelineRuns since all [Component]s are included together. This has resulted in users
+  creating multiple [Application]s that are related to each other just to resolve the visual representation
+  challenges.
+- Since a PipelineRun produces a single [Component], it becomes harder to support build strategies that produce
+  multiple artifacts, i.e [ko].
+- [Component]s are only defined in the context of an [Application] and cannot be moved. When a user wants to
+  start building a Component, they need to either add it to a current [Application] and have its builds affecting
+  the IntegrationTestScenarios results of the generated Snapshots or the component needs to be created in its
+  own temporary   Application only to be deleted and recreated in the appropriate one once the builds are
+  properly functioning.
+
+## Decision
+
+We are going to abstract the concepts of the built artifacts and their relationships to each other outside of
+any other controller within AppStudio. If a controller wants to maintain complete separation from others (i.e.
+in order to be fully deployable on its own), it may use an adapter pattern to convert this abstraction to its
+own relevant internal data model.
+
+### ArtifactConfig
+
+An `ArtifactConfig` describes the metadata which is used to define an `Artifact`, an `ArtifactStream`, and
+how `Artifact`s produced update that stream.
+
+`Artifact`s do not exist in isolation to other `Artifact`s in the system, however, there are both explicit
+and implicit relationships that exist between them, for example:
+
+- (implicit) Multiple `ArtifactConfig`s can be created for the same repository and branch combinations. These
+  configurations would be allowed as long as the name of the produced artifact is unique. This supports use
+  cases like `ko` where multiple `Artifact`s would be produced from a single build
+- (explicit) A single `ArtifactConfig` can describe the configuration for producing the "same" `Artifact"
+  from multiple branches in the same git repository. As an `ArtifactConfig` also controls the creation of
+  an `ArtifactStream`, `Artifact`s produced on any one of the branches would update the corresponding
+  `ArtifactStream`.
+- (explicit) A single repository and branch may produce different `Artifact`s under different conditions.
+  When artifacts are produced as part of a pull request, for example, the controller will still need to
+  generate `Artifact` resources, but these should not be able to update any `ArtifactStream`s for the 
+  on-push build events
+
+### Artifact
```

Comment:
How are different artifact "types" defined? I assume the `ArtifactConfig` would determine that, but can it be any type, or is it restricted to specific types (rpm, fbc, etc), which may have their own validation? (E.g. you mention some validation for RPM-based `Artifact`s farther below)